### PR TITLE
Use composer authoritative class maps in production build

### LINKED ISSUE #7337: Optimize Production Composer Build to reduce needless file lookups
### Bug Description

I recently did an optimization pass on our sites, and after profiling them found that several plugins with local composer /vendor builds were not fully optimized for production, AMP being one of these.

The plugin's local composer autoloader is not built with `--classmap-authoritative`, and by default the composer ClassLoader will prepend itself to the list of autoloaders. Both of these issues combined can cause thousands of needless file lookups for classes flowing through this loader that it doesn't need to handle. When not built with `--classmap-authoritative` and a class isn't found in the local classmap, then composer makes a few fallback attempts to find the class, resulting in all the needless file lookups.

Looks like this plugin already uses `--optimize-autoloader`, so would propose adding `--classmap-authoritative`
so adding either of these to the production build:  `-oa` or `--optimize-autoloader --classmap-authoritative`

and potentially also adding `prepend-autoloader: false` to the composer.json config, so this autoloader doesn't prepend itself ahead of other more site specific autoloaders:
```
"config": {
    "prepend-autoloader": false
}
```

### Expected Behaviour

Expected behavior is that the local composer build **only** handles autloading the classes it includes.

### Screenshots

_No response_

### PHP Version

_No response_

### Plugin Version

2.3.0

### AMP plugin template mode

Standard, Transitional, Reader

### WordPress Version

6.0.2

### Site Health

_No response_

### Gutenberg Version

_No response_

### OS(s) Affected

_No response_

### Browser(s) Affected

_No response_

### Device(s) Affected

_No response_

### Acceptance Criteria

_No response_

### Implementation Brief

_No response_

### QA Testing Instructions

_No response_

### Demo

_No response_

### Changelog Entry

_No response_

### PR DESCRIPTION
**Pull request by thelovekesh**
## Summary

Fixes #7337 

## Checklist

- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

### CODE REVIEW

**Comment by thelovekesh** [comment]
According to the docs [Authoritative class maps](https://getcomposer.org/doc/articles/autoloader-optimization.md#optimization-level-2-a-authoritative-class-maps) relieves autoloader from file lookups in case the class is not present in the class map. 

<details>
  <summary>Usage of classmap authoritative by autoloader</summary>

```php
/**
 * Finds the path to the file where the class is defined.
 *
 * @param string $class The name of the class
 *
 * @return string|false The path if found, false otherwise
 */
public function findFile($class)
{
	// class map lookup
	if (isset($this->classMap[$class])) {
		return $this->classMap[$class];
	}
	if ($this->classMapAuthoritative || isset($this->missingClasses[$class])) {
		return false;
	}
	if (null !== $this->apcuPrefix) {
		$file = apcu_fetch($this->apcuPrefix.$class, $hit);
		if ($hit) {
			return $file;
		}
	}

	$file = $this->findFileWithExtension($class, '.php');

	// Search for Hack files if we are running on HHVM
	if (false === $file && defined('HHVM_VERSION')) {
		$file = $this->findFileWithExtension($class, '.hh');
	}

	if (null !== $this->apcuPrefix) {
		apcu_add($this->apcuPrefix.$class, $file);
	}

	if (false === $file) {
		// Remember that this class does not exist.
		$this->missingClasses[$class] = true;
	}

	return $file;
}
```  
</details>

**Comment by westonruter** [comment]
I'm looking at the diff of the build before and after this change, and it seems the only change is in `vendor/composer/autoload_real.php` with the removal of these lines:

https://github.com/ampproject/amp-wp/blob/83681ec163b23d1f3c2c0c356fdee66e8739847a/vendor/composer/autoload_real.php#L37-L45

@andyblackwell Is this what you were expecting?

**Comment by andyblackwell** [comment]
> I'm looking at the diff of the build before and after this change, and it seems the only change is in `vendor/composer/autoload_real.php` with the removal of these lines:
> 
> https://github.com/ampproject/amp-wp/blob/83681ec163b23d1f3c2c0c356fdee66e8739847a/vendor/composer/autoload_real.php#L37-L45
> 
> @andyblackwell Is this what you were expecting?

Sorry, didn't see this notification. Yes, that's expected. I downloaded the production build from the github actions comment, and can see the changes you mentioned plus the line: `$loader->setClassMapAuthoritative(true);`

I also see it in the development build, which typically isn't what you'd want, but I don't know your workflow. If you don't actively make edits to the actual development build after the fact, like creating/renaming classes, then that should be fine.

**Comment by westonruter** [comment]
@andyblackwell Thank you for pointing that out.

@thelovekesh Could you revert the change to `composer_install` and instead put the change into a `composer_install_prod` command which is otherwise a copy of `composer_install`? Then in the `build` command, we can use a condition like the following:

https://github.com/ampproject/amp-wp/blob/371fa96329eef35643b4362c16f27a0f4d7fc27a/Gruntfile.js#L170-L173

To decide whether we run `composer_install` (or `composer_install_dev`) or `composer_install_prod` here:

https://github.com/ampproject/amp-wp/blob/371fa96329eef35643b4362c16f27a0f4d7fc27a/Gruntfile.js#L205

File: Gruntfile.js [hunk 1]
```diff
@@ -83,8 +83,9 @@ module.exports = function( grunt ) {
 				command: [
 					'if [ ! -e build ]; then echo "Run grunt build first."; exit 1; fi',
 					'cd build',
-					'composer install --no-dev -o',
-					'composer remove cweagans/composer-patches --update-no-dev -o',
+					// Use authoritative class maps in production by appending -oa to the composer install command.
+					`composer install --no-dev -o${ 'development' === process.env.NODE_ENV ? '' : 'a' }`,
+					`composer remove cweagans/composer-patches --update-no-dev -o${ 'development' === process.env.NODE_ENV ? '' : 'a' }`,
```
Comment (westonruter) [inline Gruntfile.js:88]: Well done!

**Review by westonruter — APPROVED** [review]

**Review by westonruter — APPROVED** [review]

**Pull request merged**